EIPS/eip-4844.md

```diff
@@ -323,33 +326,46 @@ def point_evaluation_precompile(input: Bytes) -> Bytes:
 
 ### Gas price of blobs (Simplified version)
 
-For early draft implementations, we simply change `get_blob_gas(parent)` to always return `GAS_PER_BLOB`.
+For early draft implementations, we simply change `get_blob_gas(parent)` to always return `SIMPLE_GAS_PER_BLOB`.
 
-### Gas price update rule (Full version)
+### Gas accounting (Full version)
 
-We propose a simple independent EIP-1559-style targeting rule to compute the gas cost of the transaction.
-We use the `excess_blobs` header field to store persistent data needed to compute the cost.
+We introduce data gas as a new type of gas. It is independent of normal gas and follows its own targeting rule, similar to EIP-1559.
+We use the `excess_blobs` header field to store persistent data needed to compute the data gas price. For now, only blobs are charged for in data gas. In the future this could be extended to also cover calldata (at a different relative gas cost).
 
 ```python
-def get_intrinsic_gas(tx: SignedBlobTransaction, parent: Header) -> int:
-    intrinsic_gas = 21000  # G_transaction
-    if tx.message.to == None:  # i.e. if a contract is created
-        intrinsic_gas = 53000
-    # EIP-2028 data gas cost reduction for zero bytes
-    intrinsic_gas += 16 * len(tx.message.data) - 12 * len(tx.message.data.count(0))
-    # EIP-2930 Optional access lists
-    intrinsic_gas += 1900 * sum(len(entry.storage_keys) for entry in tx.message.access_list) + 2400 * len(tx.message.access_list)
-    # New additional gas cost per blob
-    intrinsic_gas += len(tx.message.blob_versioned_hashes) * get_blob_gas(parent)
-    return intrinsic_gas
-
-def get_blob_gas(header: Header) -> int:
+def calc_data_fee(tx: SignedBlobTransaction, parent: Header) -> int:
+    return get_total_data_gas(tx) * get_data_gasprice(header)
+
+def get_total_data_gas(tx: SignedBlobTransaction) -> int:
+    return DATA_GAS_PER_BLOB * len(tx.message.blob_versioned_hashes)
+
+def get_data_gasprice(header: Header) -> int:
     return fake_exponential(
+        MIN_GASPRICE_PER_BLOB,
         header.excess_blobs,
-        GASPRICE_UPDATE_FRACTION_PER_BLOB
+        DATA_GASPRICE_UPDATE_FRACTION
     )
 ```
 
+The block validity conditions are modified to include data gas checks:
+
+```python
+def validate_block(block: Block) -> None:
+    ...
+
+    for tx in block.transactions:
+        ...
+
+        # the signer must be able to afford the transaction
+        assert signer(tx).balance >= tx.message.gas * tx.message.max_fee_per_gas + get_total_data_gas(tx) * tx.message.max_fee_per_data_gas
+
+        # ensure that the user was willing to at least pay the current data gasprice
+        assert tx.message.max_fee_per_data_gas >= get_data_gasprice(parent(block).header)
+```
+
+The actual `data_fee` as calculated via `calc_data_fee` is deducted from the sender balance before transaction execution, and is not refunded in case of transaction failure.
```

Comment:
Makes sense to not have the data fee be refunded, given that the data persists even on reverted transaction.

Reply:
And can be operated on in a subsequent normal TX with a sufficient merkle proof into a prior header and TX list...

That said, it might make sense to put some sort of accumulator of KZG commitments in the EL or CL state so that they can be operated on easily in a historic fashion. This would be easy on CL and accessible to EL if 4788 ships